dvc/repo/run.py

```diff
@@ -109,12 +110,12 @@ def run(self, fname=None, no_exec=False, single_stage=False, **kwargs):
         return None
 
     dvcfile = Dvcfile(self, stage.path)
-    if kwargs.get("force", True):
-        dvcfile.remove_stage(stage)
-    else:
-        _check_stage_exists(dvcfile, stage)
-
     try:
+        if kwargs.get("force", True):
+            with suppress(ValueError):
+                self.stages.remove(stage)
+        else:
+            _check_stage_exists(dvcfile, stage)
```

Comment:
So, a bit more details on the implementation:

We check the graph for duplications. But, `check_modified_graph` appends the newly created `stage` and `self.stages` (which might contain the stage that we are trying to overwrite). For `-f`, we will be overwriting it anyway, so we can do
graph checks by removing the stage from self.stages and later appending the new stage.

 For `dvc.yaml`, we do a diff later when dumping, so, this preserves comments and meta.
For `.dvc` files, the `dvc run` never loads the file and therefore, does not set `_stage_text`, and hence, the `dump` of single stage file is never aware of `meta` and comments present before, so, it just overwrites them.

Reply:
This process is however slow, because there's clearly a `O(n)` for going throught the list of `self.stages`, and equality comparisons might not be faster (considering the findings with `relpath` :smile:).

Note, the equality of stages for `dvc.yaml` stages is the path relative to root of the repo and name of the stage whereas for `.dvc` files, it's just path rel. to the root.